tsfile/src/main/java/org/apache/iotdb/tsfile/encoding/encoder/RegularDataEncoder.java

```diff
@@ -0,0 +1,438 @@
+      // get the new regular data if the missing point exists in the original data
+      data = getRegularData(data);
+      firstValue = data[0];
+      if (isMissingPoint) {
+        writeIndex = data.length;
+      }
+      if (!isLastPack) {
+        flush(out);
+      }
+    }
+
+    private void writeBitmap(ByteArrayOutputStream out) throws IOException {
+      // generate bitmap
+      data2Diff(missingPointData);
+      byte[] bsArr = bitmap.toByteArray();
+      out.write(BytesUtils.intToBytes(bsArr.length));
+      out.write(bsArr);
+    }
+
+    private int[] getRegularData(int[] data) {
+      if (writeIndex > 1) {
+        previousValue = data[0];
+        minDeltaBase = data[1] - data[0];
+        // calculate minimum elapsed of the data and check whether the missing point exists
+        for (int i = 1; i < writeIndex; i++) {
+          calcDelta(data[i]);
+          previousValue = data[i];
+        }
+      }
+      // generate the continious data when the missing point exists
+      if (isMissingPoint) {
+        generateRegularData(data);
+        return regularData;
+      }
+      return data;
+    }
+
+    private void generateRegularData(int[] data) {
+      missingPointData = data;
+      blockSize = (int) (((data[writeIndex - 1] - data[0]) / minDeltaBase) + 1);
+      regularData = new int[blockSize];
+      for (int i = 0; i < blockSize; i++) {
+        regularData[i] = (data[0] + minDeltaBase * i);
+      }
+    }
+
+    private void calcDelta(int value) {
+      int delta = value - previousValue; // calculate delta
+      if (delta != minDeltaBase) {
+        isMissingPoint = true;
+      }
+      if (delta < minDeltaBase) {
+        minDeltaBase = delta;
+      }
+    }
+
+    private void data2Diff(int[] missingPointData) {
+      bitmap = new BitSet(regularData.length);
+      bitmap.flip(0, regularData.length);
+      int offset = 0;
+      for (int i = 1; i < missingPointData.length; i++) {
+        long delta = missingPointData[i] - missingPointData[i - 1];
+        if (delta != minDeltaBase) {
+          int missingPointNum = (int) (delta / minDeltaBase) - 1;
+          for (int j = 0; j < missingPointNum; j++) {
+            bitmap.set(i + (offset++), false);
+          }
+        }
+      }
+    }
+  }
+
+  public static class LongRegularEncoder extends RegularDataEncoder {
+
+    private long[] data;
+    private long[] missingPointData;
+    private long[] regularData;
+    private long firstValue;
+    private long previousValue;
+    private long minDeltaBase;
+    private boolean isMissingPoint;
+    private boolean isLastPack;
+    private BitSet bitmap;
+
+    public LongRegularEncoder() {
+      this(BLOCK_DEFAULT_SIZE);
+    }
+
+    /**
+     * constructor of LongRegularEncoder which is a sub-class of RegularDataEncoder.
+     *
+     * @param size - the number how many numbers to be packed into a block.
+     */
+    public LongRegularEncoder(int size) {
```

Comment:
Refer to all comments in `IntRegularEncoder`.

Reply:
Done. Thx.